examples/remix-basic/app/routes/_index.tsx

```diff
@@ -0,0 +1,101 @@
+import { useShape, preloadShape, getShapeStream } from "@electric-sql/react"
+import { useFetchers, Form } from "@remix-run/react"
+import { v4 as uuidv4 } from "uuid"
+import type { ClientActionFunctionArgs } from "@remix-run/react"
+import "../Example.css"
+import { matchStream } from "../match-stream"
+
+const itemShape = () => {
+  return {
+    url: new URL(`/shape-proxy/items`, window.location.origin).href,
+  }
+}
+
+export const clientLoader = async () => {
+  return await preloadShape(itemShape())
+}
+
+export const clientAction = async ({ request }: ClientActionFunctionArgs) => {
+  const body = await request.formData()
+
+  const itemsStream = getShapeStream(itemShape())
+
+  if (body.get(`intent`) === `add`) {
+    // Match the insert
+    const findUpdatePromise = matchStream({
+      stream: itemsStream,
+      operations: [`insert`],
+      matchFn: ({ message }) => message.value.id === body.get(`new-id`),
+    })
+
+    // Generate new UUID and post to backend
+    const fetchPromise = fetch(`/api/items`, {
+      method: `POST`,
+      body: JSON.stringify({ uuid: body.get(`new-id`) }),
+    })
+
+    return await Promise.all([findUpdatePromise, fetchPromise])
+  } else if (body.get(`intent`) === `clear`) {
+    // Match the delete
+    const findUpdatePromise = matchStream({
+      stream: itemsStream,
+      operations: [`delete`],
+      // First delete will match
+      matchFn: () => true,
+    })
+    // Post to backend to delete everything
+    const fetchPromise = fetch(`/api/items`, {
+      method: `DELETE`,
+    })
+
+    return await Promise.all([findUpdatePromise, fetchPromise])
+  }
+}
+
+type Item = { id: string }
+
+export default function Example() {
+  const { data: items } = useShape(itemShape()) as unknown as { data: Item[] }
+
+  const submissions = useFetchers()
+    .filter((fetcher) => fetcher.formData?.get(`intent`) === `add`)
+    .map((fetcher) => {
+      return { id: fetcher.formData?.get(`new-id`) } as Item
+    })
+
+  const isClearing = useFetchers().some(
+    (fetcher) => fetcher.formData?.get(`intent`) === `clear`
+  )
+
+  // Merge data from shape & optimistic data from fetchers. This removes
+  // possible duplicates as there's a potential race condition where
+  // useShape updates from the stream slightly before the action has finished.
+  const itemsMap = new Map()
+  items.concat(submissions).forEach((item) => {
```

Comment:
I understand you followed a minimal approach for rebasing the pending state and I don't have any objections to the approach.

My observation is that the code for rebasing is not trivial to write but is kinda systematic. So eventually we could wrap it in a useTentativeShape that would receive the state from `useShape` and a tentative state from `submissions` and do the rebasing internally. I think this would make the example code more reusable too.

Reply:
Yes if we identify common patterns then rolling it into the library would make sense. I'm not 100% sure we will though because people use a lot of patterns for forms & mutations and how they want to show optimistic state e.g. here I used `new-id` from the form with the new UUID.

I think leaving messy patterns as messy is fine. It's not that hard I suspect for people can figure it out for their own projects for now. Probably there'll be libraries that emerge around specific patterns for specific frameworks.

Reply:
Keeping our footprint really really small helps us move fast early on. So unless we're sure most people would want something, I think we should just do examples vs. library code. E.g. the merging logic here is specific to this example but is easy to generalize. But anyone doing optimistic state will want some sort of `matchState` function.